### PR TITLE
feat(scaffolder): create custom action for scaffolder templates

### PR DESCRIPTION
**Pull request by debsmita1**
**Resolves:**

https://issues.redhat.com/browse/RHIDP-2192
[RHIDP-2221](https://issues.redhat.com/browse/RHIDP-2221)

**Test setup:**

- Follow the README.md to integrate the custom action plugin in the backend.

- Changes in `app-config.yaml`

```
catalog:
  location:
    # Example template
    - type: url
      target: https://github.com/debsmita1/nodejs-ex-1/blob/master/local-template.yaml
      rules:
        - allow: [Template]
```

- Changes in `App.tsx`

```
import {
  CatalogEntityPage,
  CatalogIndexPage,
  CatalogTable,
  CatalogTableRow,
  CatalogTableColumnsFunc,
} from '@backstage/plugin-catalog';


export const AppBase = () => {
  const { AppProvider, AppRouter, dynamicRoutes } =
    useContext(DynamicRootContext);
  const myCustomColumnsFunc: CatalogTableColumnsFunc = entityListContext => {
    if (entityListContext.filters.kind?.value === 'component') {
      return [
        ...CatalogTable.defaultColumnsFunc(entityListContext),
        {
          title: 'Created At',
          ender: (data: CatalogTableRow) =>
            data.entity.metadata.annotations?.['backstage.io/createdAt'],
        },
      ];
    }

    return CatalogTable.defaultColumnsFunc(entityListContext);
  };
  .....
  
  
  <Route
       path="/catalog"
       element={<CatalogIndexPage columns={myCustomColumnsFunc} />}
   />
```

### CODE REVIEW

**Review by rohitkrai03 — COMMENTED** [review]
Overall looks good to me but I am thinking if we really need a custom action for this. Are there plans to use this custom action for other use cases as well?

**Comment by debsmita1** [comment]
> Overall looks good to me but I am thinking if we really need a custom action for this. Are there plans to use this custom action for other use cases as well?

@rohitkrai03 updated the PR to create a generic action to annotate objects with labels and annotations. PTAL

**Comment by Zaperex** [comment]
@debsmita1 I noticed you added a generic custom action for annotating the labels and annotations. I have a similar [custom action](https://github.com/janus-idp/backstage-plugins/pull/1605) in the works that appends a `spec.scaffoldedFrom` field to the `catalog-info.yaml` to enable a custom relation link between software templates and the catalog entities they generated. 

I was wondering if it would be a good idea to extend this generic action to allow modification of the spec field (and potentially other fields in the future). Then we can potentially export and publish both of these `catalog-info.yaml` modification custom actions (timestamping, adding `templatedFrom` field) alongside  the generic custom action as a common package of `catalog-info.yaml` modification custom actions similar to how the [built-in backstage actions](https://github.com/backstage/backstage/blob/master/plugins/scaffolder-backend/src/scaffolder/actions/builtin/createBuiltinActions.ts) are exported rather than having 2 separate custom action packages published?

**Comment by debsmita1** [comment]
> @debsmita1 I noticed you added a generic custom action for annotating the labels and annotations. I have a similar [custom action](https://github.com/janus-idp/backstage-plugins/pull/1605) in the works that appends a `spec.scaffoldedFrom` field to the `catalog-info.yaml` to enable a custom relation link between software templates and the catalog entities they generated.
> 
> I was wondering if it would be a good idea to extend this generic action to allow modification of the spec field (and potentially other fields in the future). Then we can potentially export and publish both of these `catalog-info.yaml` modification custom actions (timestamping, adding `spec.templatedFrom` field) alongside this generic custom action as a common package of `catalog-info.yaml` modification custom actions rather than publishing multiple packages for all the future `catalog-info.yaml` modification custom actions. This would be similar to how the [built-in backstage actions](https://github.com/backstage/backstage/blob/master/plugins/scaffolder-backend/src/scaffolder/actions/builtin/createBuiltinActions.ts) are published and exported.
> 
> What are you thoughts?

Absolutely! This totally makes sense. I will update my PR to allow modification in the `spec` field

**Comment by Zaperex** [comment]
I made a PR (https://github.com/debsmita1/backstage-plugins/pull/3) on your branch with some fixes to the `package.json` exporting to allow it to work in the new backend system. Also fixed up the `yarn lint` issues and the `yarn build` issues so it should pass the tests.

File: plugins/scaffolder-annotator-action/README.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  import('@janus-idp/backstage-scaffolder-backend-module-annotator/alpha'),
+);
+// highlight-add-end
+
+backend.start();
+```
+
+### Installing on the legacy backend system
+
+1. Install the Annotator custom action plugin using the following command:
```
Comment (rohitkrai03) [inline plugins/scaffolder-annotator-action/README.md:25]: Maybe add a section on top that explains what this custom action package provides overall? Like all the features and then the specific actions for timestamping and scafolded-from.

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]
@debsmita1 Overall looks good. I think it would be beneficial if you could also provide an example template, similar to the one shown here - https://github.com/janus-idp/backstage-plugins/tree/main/plugins/kubernetes-actions/examples/templates.

**Comment by ciiay** [comment]
Hi @debsmita1 , I have tried this PR locally and I got "Template action with ID 'catalog:timestamping' is not registered." error. 
rhidp_2192
Did I miss anything?

**Comment by debsmita1** [comment]
> Hi @debsmita1 , I have tried this PR locally and I got "Template action with ID 'catalog:timestamping' is not registered." error. rhidp_2192 Did I miss anything?

Thanks @ciiay for trying it out. Do you have the new changes in the `scaffolder.ts` file as mentioned in the Readme ?

**Comment by ciiay** [comment]
Tested in backstage-showcase and it works great 👍 
image

**Comment by invincibleJai** [comment]
/lgtm

Thanks @debsmita1, this works as described. I tested as static and dynamic way in showcase

File: plugins/scaffolder-annotator-action/README.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+```console
+yarn workspace backend add @janus-idp/backstage-scaffolder-backend-module-annotator
+```
+
+2. Integrate the custom actions in the `scaffolder-backend` `createRouter` function:
```
Comment (hmanwani-rh) [inline plugins/scaffolder-annotator-action/README.md:33]:
```suggestion
2. Integrate the custom actions in the `scaffolder-backend` `createRouter` function using the following steps:
```

File: plugins/scaffolder-annotator-action/README.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+  return await createRouter({
+    /* highlight-add-next-line */
+    actions,
+    logger: env.logger,
+    ...
+  });
+  ```
+
+3. To annotate the catalog-info.yaml skeleton with the current timestamp, add the **catalog:timestamping** custom action in your template yaml after the `Fetch Skeleton + Template` step:
```
Comment (hmanwani-rh) [inline plugins/scaffolder-annotator-action/README.md:82]:
```suggestion
3. To annotate the catalog-info.yaml skeleton with the current timestamp, add the `catalog:timestamping` custom action in your template YAML file after the `Fetch Skeleton + Template` step:
```

File: plugins/scaffolder-annotator-action/README.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    ...
+
+  # highlight-add-start
+  - id: timestamp
+    name: Add Timestamp to catalog-info.yaml
+    action: catalog:timestamping
+  # highlight-add-end
+```
+
+4. To annotate the catalog-info.yaml skeleton with the template entityRef, add the **catalog:scaffolded-from** custom action in your template yaml after the `Fetch Skeleton + Template` step:
```
Comment (hmanwani-rh) [inline plugins/scaffolder-annotator-action/README.md:98]:
```suggestion
4. To annotate the `catalog-info.yaml` skeleton with the template `entityRef`, add the `catalog:scaffolded-from` custom action in your YAML file after the `Fetch Skeleton + Template` step:
```

File: plugins/scaffolder-annotator-action/README.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  - id: append-templateRef
+    name: Append the entityRef of this template to the entityRef
+    action: catalog:scaffolded-from
+  # highlight-add-end
+
+```
+
+## Usage
+
+To use the `createAnnotatorAction` to create a custom action
```
Comment (hmanwani-rh) [inline plugins/scaffolder-annotator-action/README.md:116]:
```suggestion
You can use the `createAnnotatorAction` to create a custom action using the following parameters:
```

File: plugins/scaffolder-annotator-action/README.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+| `annotateEntityObject.labels`      | `object` |    No    | Key-value pairs to be added to the `metadata.labels` of the entity                                                                                                                                                                                       |
+| `annotateEntityObject.annotations` | `object` |    No    | Key-value pairs to be added to the `metadata.annotations` of the entity                                                                                                                                                                                  |
+| `annotateEntityObject.spec`        | `object` |    No    | Key-value pairs to be added to the `metadata.spec` of the entity. The value for each key can either be a string or an object with the key `readFromContext`, enabling users to specify the path in the context from which the value should be retrieved. |
+
+The annotator action accepts the following inputs
```
Comment (hmanwani-rh) [inline plugins/scaffolder-annotator-action/README.md:127]:
```suggestion
The annotator action plugin accepts the following inputs:
```

**Review by hmanwani-rh — COMMENTED** [review]
@debsmita1 Added a few comments. Please take a look and let me know if you've any questions.

**Pull request merged**